### PR TITLE
Fixed using a multi repository medium (bsc#1062813)

### PR DESCRIPTION
**Pull request by lslezak**
- All new added repositories are now processed, not only one of them
- See https://bugzilla.suse.com/show_bug.cgi?id=1062813
- Added initial unit test support
- Added some simple unit tests
- Enabled Coveralls (code coverage)
- Also tested manually in a full installation (see the screenshots below)
- Some additional cleanup

## Screenshots

If user selected several addons from a multi repository medium in this dialog:

multi_product_selection

Then the addon summary listed only one of them:

multi_product_list_broken

And only that product was selected to install:

multi_product_summary_broken

## With the Fix

After applying the fix all addons are displayed in the table:

multi_product_list

And all products are selected to install:

multi_product_summary

### CODE REVIEW

File: src/include/add-on/add-on-workflow.rb [hunk 1]
```diff
@@ -129,44 +134,32 @@ def GetURLBeginsWith(source_type)
...
+      if src_id.nil?
+        log.error("Wrong src_id: #{src_id}")
         return
       end
 
       # BNC #450274
       # Prevent from adding one product twice
-      matching_products = Builtins.filter(AddOnProduct.add_on_products) do |one_product|
-        Ops.get_integer(one_product, "media", -1) == src_id
+      matching_product = AddOnProduct.add_on_products.find do |one_product|
```
Comment (jreidinger) [inline src/include/add-on/add-on-workflow.rb]: I would name it simple `product`. one_product sounds strange.

File: src/include/add-on/add-on-workflow.rb [hunk 2]
```diff
@@ -224,37 +217,31 @@ def MediaSelect
         sources_after = Pkg.SourceGetCurrent(false)
         Builtins.y2milestone("Sources with new one added: %1", sources_after)
 
-        # bnc #393011
-        # AddOnProduct::src_id must be set to the latest source ID
-        src_id_found = false
+        @added_repos = []
```
Comment (jreidinger) [inline src/include/add-on/add-on-workflow.rb]: this method is already long as a week before payment. Is possible to move this `:next` handling to own method?

File: src/include/add-on/add-on-workflow.rb [hunk 3]
```diff
@@ -224,37 +217,31 @@ def MediaSelect
...
 
-        Builtins.foreach(sources_after) do |one_source|
-          if !Builtins.contains(sources_before, one_source)
-            AddOnProduct.src_id = one_source
-            Builtins.y2milestone("Added source ID is: %1", AddOnProduct.src_id)
-            src_id_found = true
-            raise Break
-          end
+        sources_after.each do |source|
+          next if sources_before.include?(source)
```
Comment (jreidinger) [inline src/include/add-on/add-on-workflow.rb]: I think this deserve a comment as it is not so clear. Is it mean that sources that are already in system is skipped? So it is not mentioned in newly added sources?

File: src/include/add-on/add-on-workflow.rb [hunk 4]
```diff
@@ -224,37 +217,31 @@ def MediaSelect
...
+          end
         else
-          AddOnProduct.src_id = Ops.get(
-            sources_after,
-            Ops.subtract(Builtins.size(sources_after), 1),
-            0
-          )
-          Builtins.y2warning("Fallback src_id: %1", AddOnProduct.src_id)
+          @added_repos << sources_after.last
+          log.warn("Fallback src_id: #{sources_after.last}")
```
Comment (jreidinger) [inline src/include/add-on/add-on-workflow.rb]: I have no idea what is going here. Can you please add comment explaining this?

File: src/include/add-on/add-on-workflow.rb [hunk 5]
```diff
@@ -458,49 +445,23 @@ def CatalogSelect
     # BNC #474745
     # Installs all the products from just added add-on media
     def InstallProduct
-      Builtins.y2milestone("AddOnID: %1", AddOnProduct.src_id)
+      log.info("Installing products from #{@added_repos.inspect}")
```
Comment (jreidinger) [inline src/include/add-on/add-on-workflow.rb]: please document that this method always return `:next` in yardoc

File: src/include/add-on/add-on-workflow.rb [hunk 6]
```diff
@@ -458,49 +445,23 @@ def CatalogSelect
...
-        #	    y2warning ("Product %1 is already installed", this_product);
-        #	    return;
-        #	}
-        product_name = Ops.get_string(one_product, "name", "-Unknown-Product-")
-        Builtins.y2milestone(
-          "Selecting product '%1' for installation -> %2",
-          product_name,
-          Pkg.ResolvableInstall(product_name, :product)
-        )
+        next if product["status"] != :available
```
Comment (jreidinger) [inline src/include/add-on/add-on-workflow.rb]: I think this deserve some log entry when adding product that is not avaialble.

File: src/include/add-on/add-on-workflow.rb [hunk 7]
```diff
@@ -890,9 +837,10 @@ def RunWizard
     end
 
     def RunAutorunWizard
-      aliases = { "catalog" => lambda { CatalogSelect() }, "product" => lambda do
-        ProductSelect()
-      end }
+      aliases = {
+        "catalog" => lambda { CatalogSelect() },
+        "product" => lambda { ProductSelect() }
+      }
```
Comment (jreidinger) [inline src/include/add-on/add-on-workflow.rb]: :+1:

File: src/include/add-on/add-on-workflow.rb [hunk 8]
```diff
@@ -1175,32 +1121,29 @@ def RunAddOnMainDialog(enable_back, enable_next, enable_abort, back_button, next
 
           if ret2 == :next
             # Add-On product has been added, integrate it (change workflow, use y2update)
-            AddOnProduct.Integrate(AddOnProduct.src_id)
+            @added_repos.each { |src_id| AddOnProduct.Integrate(src_id) }
 
             # check whether it requests registration (FATE #301312)
-            AddOnProduct.PrepareForRegistration(AddOnProduct.src_id)
+            @added_repos.each { |src_id| AddOnProduct.PrepareForRegistration(src_id) }
```
Comment (jreidinger) [inline src/include/add-on/add-on-workflow.rb]: can it be done in one step?
like
```
@added_repos.each do |src_id|
  AddOnProduct.Integrate(src_id)
  AddOnProduct.PrepareForRegistration(src_id)
end
```

File: src/include/add-on/add-on-workflow.rb [hunk 9]
```diff
@@ -1175,32 +1121,29 @@ def RunAddOnMainDialog(enable_back, enable_next, enable_abort, back_button, next
...
             ret = nil if ret == :skip_to_add
           elsif ret2 == :abort || ret2 == :cancel
             Builtins.y2milestone("Add-on sequence aborted")
 
-            if AddOnProduct.src_id != nil
-              Builtins.y2milestone(
-                "Removing add-on repository: %1",
-                AddOnProduct.src_id
-              )
+            if !@added_repos.empty?
```
Comment (jreidinger) [inline src/include/add-on/add-on-workflow.rb]: I think it is not needed to check it. each will work fine with empty array and log that it removing empty repositories looks consistent for me.

File: src/include/add-on/add-on-workflow.rb (hunk 4 above)
Comment (lslezak) [reply src/include/add-on/add-on-workflow.rb]: Hmm, no idea, IMHO this should never happen. I just added a FIXME note, I do not want to break something...

File: src/include/add-on/add-on-workflow.rb (hunk 8 above)
Comment (lslezak) [reply src/include/add-on/add-on-workflow.rb]: I'm not sure, the methods call quite a lot of other functions, I just added a comment to possibly fix that later.

**Pull request merged**